mshadow/stream_gpu-inl.h

```diff
@@ -127,17 +134,29 @@ struct Stream<gpu> {
     this->DestroyDnnHandle();
     cudnnStatus_t err = cudnnCreate(&dnn_handle_);
     CHECK_EQ(err, CUDNN_STATUS_SUCCESS) << cudnnGetErrorString(err);
+    // At this point, we have the resource which may need to be freed
+    this->dnn_handle_ownership_ = OwnHandle;
     err = cudnnSetStream(dnn_handle_, stream_);
     CHECK_EQ(err, CUDNN_STATUS_SUCCESS) << cudnnGetErrorString(err);
-    this->dnn_handle_ownership_ = OwnHandle;
 #endif
   }
 };
 template<>
+inline void DeleteStream<gpu>(Stream<gpu> *stream) {
+  if(stream) {
+    MSHADOW_CUDA_CALL(cudaStreamDestroy(stream->stream_));
+    stream->DestoryBlasHandle();
+    stream->DestroyDnnHandle();
+    delete stream;
+  }
+}
+template<>
 inline Stream<gpu> *NewStream<gpu>(bool create_blas_handle,
                                    bool create_dnn_handle,
                                    int dev_id) {
-  Stream<gpu> *st = new Stream<gpu>();
+  // RAII on Cuda exception
+  struct StreamDeleter { void operator()(Stream<gpu> *ptr) const { DeleteStream<gpu>(ptr); } };
+  std::unique_ptr<Stream<gpu>, StreamDeleter> st(new Stream<gpu>());
```

Comment:
In case cuda exception is thrown, memory and any allocate handles are automatically cleaned up via RAII